### PR TITLE
feat: create max-nested-describe rule

### LINKED ISSUE #362: Support `max-describe-nesting-level` rule.
Currently we don't have any rule that could control level of `describe`'s nesting.

```
rules: {
   "jest/max-describe-nesting-level": ["warn", {limit: 1}]
},
```

Good
```
describe('getSomething', function () {
    it('should get green plastic chair with `green` argument', () => {
      expect(getSomething('green').toBe('Green plastic chair'))
    });
});
```

Bad
```
describe('getSomething', function () {
    describe('w/o arguments', function () {
      it('should get orange jacket', () => {
        expect(getSomething().toBe('Orange jacket'))
      });
    })
});
```

I can implement this (with better examples 👀) if you are good with this rule?

### PR DESCRIPTION
**Pull request by dominictwlee**
Resolves #362

The idea is the same as the above issue. I changed the rule name and options to something more consistent with the `max-nested-callbacks` rule.

### CODE REVIEW

File: src/rules/utils.ts [hunk 1]
```diff
@@ -648,6 +648,16 @@ export type FunctionExpression =
   | TSESTree.ArrowFunctionExpression
   | TSESTree.FunctionExpression;
 
+export const isCallExpression = (
```
Comment (G-Rath) [inline src/rules/utils.ts:651]: We don't need this utility function, as you can just do `node?.type === AST_NODE_TYPES.CallExpression`

File: src/rules/max-nested-describe.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    },
+    messages: {
+      exceededMaxDepth:
+        'Too many nested describe calls ({{depth}}). Maximum allowed is {{max}}.',
+    },
+    type: 'suggestion',
+    schema: [
+      {
+        oneOf: [
+          {
```
Comment (G-Rath) [inline src/rules/max-nested-describe.ts:20]: Let's keep things simple and just support an object

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import rule from '../max-nested-describe';
+
+const ruleTester = new TSESLint.RuleTester({
+  parser: resolveFrom(require.resolve('eslint'), 'espree'),
+  parserOptions: {
+    ecmaVersion: 2017,
+  },
+});
+
+ruleTester.run('max-nested-describe', rule, {
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:13]: Could you add a test for `{ max: 0 }`

**Review by G-Rath — CHANGES_REQUESTED** [review]

**Comment by dominictwlee** [comment]
Thanks for the review @G-Rath, I've made the changes according to your comments.

File: src/rules/__tests__/max-nested-describe.test.ts (hunk 3 above)
Comment (G-Rath) [reply src/rules/__tests__/max-nested-describe.test.ts:13]: Could you chuck in a few tests for prefixed & modified describes? (i.e `fdescribe`, `describe.skip` - feel free to just chuck these into some of the existing tests), and one of each version of `describe.each`.

File: docs/rules/max-nested-describe.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+```
+
+## Options
+
+```json
+{
+  "jest/max-nested-describe": [
+    "error",
+    {
+      "max": 3
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:69]:
```suggestion
      "max": 2
```

nit: I think it's better if we use the default value here.

File: docs/rules/max-nested-describe.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+```
+
+The following patterns are **not** considered warnings (with the default option
+of `{ "max": 2 } `):
+
+```js
+describe('foo', () => {
+  describe('bar', () => {
+    it('should get something', () => {
+      expect(getSomething().toBe('Something'));
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:44]:
```suggestion
      expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    });
+  });
+
+  describe('qux', () => {
+    it('should get something', () => {
+      expect(getSomething().toBe('Something'));
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:50]:
```suggestion
      expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    });
+  });
+});
+
+describe('foo2', function () {
+  it('should get something', () => {
+    expect(getSomething().toBe('Something'));
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:57]:
```suggestion
    expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      });
+    });
+  });
+});
+
+describe('foo2', function () {
+  describe('bar2', function () {
+    describe('baz2', function () {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'));
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:30]:
```suggestion
        expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+The following patterns are considered warnings (with the default option of
+`{ "max": 2 } `):
+
+```js
+describe('foo', () => {
+  describe('bar', () => {
+    describe('baz', () => {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'));
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:20]:
```suggestion
        expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+Examples of patterns **not** considered warnings with options set to
+`{ "max": 3 }`:
+
+```js
+describe('foo', () => {
+  describe('bar', () => {
+    describe('baz', () => {
+      it('should get something', () => {
+        expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:89]:
```suggestion
        expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+      });
+    });
+  });
+});
+
+describe('foo2', function()) {
+  describe('bar2', function() {
+    describe('baz2', function() {
+      it('should get something', function() {
+        expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:99]:
```suggestion
        expect(getSomething()).toBe('Something');
```

File: docs/rules/max-nested-describe.md [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+      });
+    });
+
+    describe('qux2', function() {
+      it('should get something', function() {
+        expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline docs/rules/max-nested-describe.md:105]:
```suggestion
        expect(getSomething()).toBe('Something');
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
Cheers, this is looking good! Just got some documentation fixes, and if you could add a few modifier & `.each` tests that'd be great!

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  ],
+  invalid: [
+    {
+      code: dedent`
+        describe('foo', function() {
+          describe('bar', function () {
+            describe('baz', function () {
+              describe('qux', function () {
+                it('should get something', () => {
+                  expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:100]:
```suggestion
                  expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+        { messageId: 'exceededMaxDepth', line: 4, column: 7 },
+      ],
+    },
+    {
+      code: dedent`
+        describe('foo', () => {
+          describe('bar', () => {
+            describe('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:118]:
```suggestion
                expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+              });
+            });
+
+            describe('qux', function () {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:124]:
```suggestion
                expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+        { messageId: 'exceededMaxDepth', line: 9, column: 5 },
+      ],
+    },
+    {
+      code: dedent`
+        fdescribe('foo', () => {
+          describe.only('bar', () => {
+            describe.skip('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:141]:
```suggestion
                expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+              });
+            });
+
+            describe('baz', () => {
+              it('should get something', () => {
+                expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:147]:
```suggestion
                expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+              });
+            });
+          });
+        });
+
+        xdescribe('qux', () => {
+          it('should get something', () => {
+            expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:155]:
```suggestion
            expect(getSomething()).toBe('Something');
```

File: src/rules/__tests__/max-nested-describe.test.ts [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+      errors: [
+        { messageId: 'exceededMaxDepth', line: 3, column: 5 },
+        { messageId: 'exceededMaxDepth', line: 9, column: 5 },
+      ],
+    },
+    {
+      code: dedent`
+        describe('qux', () => {
+          it('should get something', () => {
+            expect(getSomething().toBe('Something'))
```
Comment (G-Rath) [inline src/rules/__tests__/max-nested-describe.test.ts:168]:
```suggestion
            expect(getSomething()).toBe('Something');
```

**Comment by dominictwlee** [comment]
My bad with the brackets! been a long day my eyes must be going fuzzy.

File: src/rules/max-nested-describe.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Enforces a maximum depth to nested describe calls',
+      recommended: false,
+    },
+    messages: {
+      exceededMaxDepth:
+        'Too many nested describe calls ({{depth}}). Maximum allowed is {{max}}.',
```
Comment (G-Rath) [inline src/rules/max-nested-describe.ts:17]:
```suggestion
        'Too many nested describe calls ({{ depth }}). Maximum allowed is {{ max }}.',
```

nit: I typically put spaces around the names (and so most if not all of our rules are like this) - we don't have a lint rule for enforcing this, but would be nice to be consistent.

**Review by G-Rath — APPROVED** [review]
Awesome! No worries about the brackets - thanks for putting up with all my nitty requests 😄

**Comment by G-Rath** [comment]
I'm now wondering if `2` is a good max, given that `max-nested-callbacks` has a max of `10`.

Thinking about the tests I've written, I easily have more than two nested `describe`s, and think you would easily hit that with class/method tests:

```
describe('MyClass', () => {
  describe('#doSomethingWithDate', () => {
    describe('when an error happens', () => {
       ...
    });

    describe('when given a number', () => {
      it('parses it as a date', () => {
         ...
      });
    });

    describe('when the date is in the future', () => {
       ...
    });
  });
});
```

I'm not overly fussed about having the perfect default, but wouldn't mind hearing if @SimenB has any opinions on what a good nested level might be.

**Comment by dominictwlee** [comment]
> I'm now wondering if `2` is a good default max, given that `max-nested-callbacks` has a max of `10`.
> 
> Thinking about the tests I've written, I easily have more than two nested `describe`s, and think you would easily hit that with class/method tests:
> 
> ```
> describe('MyClass', () => {
>   describe('#doSomethingWithDate', () => {
>     describe('when an error happens', () => {
>        ...
>     });
> 
>     describe('when given a number', () => {
>       it('parses it as a date', () => {
>          ...
>       });
>     });
> 
>     describe('when the date is in the future', () => {
>        ...
>     });
>   });
> });
> ```
> 
> I'm not overly fussed about having the perfect default, but wouldn't mind hearing if @SimenB has any opinions on what a good nested level might be.

Yeah it's pretty subjective so probably depends on the respective individuals/teams.Though I do think 10 is probably a bit much for nested describes.

**Comment by dominictwlee** [comment]
@SimenB @G-Rath Hey just following up on this. Any thoughts on the dafaults?

**Comment by G-Rath** [comment]
@dominictwlee sorry I've been very busy recently, both working on some core changes to `eslint-plugin-jest` (currently, looking at using `settings` for configuring rules) and with other work stuff 😅

Currently I think 5 is probably a good default, as it leaves you with 3 levels of nesting after a class describe + method describe.

**Comment by dominictwlee** [comment]
> @dominictwlee sorry I've been very busy recently, both working on some core changes to `eslint-plugin-jest` (currently, looking at using `settings` for configuring rules) and with other work stuff 
> 
> Currently I think 5 is probably a good default, as it leaves you with 3 levels of nesting after a class describe + method describe.

@G-Rath  I've updated the default max to 5, and added a few tests + examples in docs for this.

**Comment by dominictwlee** [comment]
@G-Rath Just following up on this periodically, let me know if anything needs to change.

**Comment by G-Rath** [comment]
@dominictwlee cheers, lets land it

**Pull request merged**